src/metatracer.rs

```diff
@@ -30,57 +30,89 @@
+    }
 }
 
 /// A meta-tracer.
-pub struct MetaTracer<Loc> {
-    counts: Mutex<HashMap<Loc, LocPhase>>
+pub struct MetaTracer {
+    hot_threshold: AtomicU32
 }
 
-impl<Loc: Eq + Hash> MetaTracer<Loc> {
+impl MetaTracer {
     pub fn new() -> Self {
         Self {
-            counts: Mutex::new(HashMap::new())
+            hot_threshold: AtomicU32::new(DEFAULT_HOT_THRESHOLD)
         }
     }
 
+    /// Update the hot threshold for this meta-tracer instance to `threshold`.
+    pub fn set_hot_threshold(&mut self, threshold: HotThreshold) {
+        self.hot_threshold.store(threshold, Ordering::Relaxed);
+    }
+
     /// Attempt to execute a compiled trace for location `loc`: return a `ControlOutcome` to allow
     /// the end user to determine whether a trace was executed or not.
-    pub fn control_point(&self, loc: Loc) {
-        let mut cnts = self.counts.lock().unwrap();
-        match cnts.entry(loc) {
-            Entry::Occupied(mut v) => {
-                match *v.get() {
-                    LocPhase::Counting(x) => {
-                        if x >= HOT_THRESHOLD {
-                            v.insert(LocPhase::Tracing);
-                        } else {
-                            v.insert(LocPhase::Counting(x + 1));
-                        }
-                    },
-                    LocPhase::Tracing => {
-                        v.insert(LocPhase::Compiled);
-                    },
-                    LocPhase::Compiled => ()
-                }
-            },
-            Entry::Vacant(v) => {
-                v.insert(LocPhase::Counting(1));
+    pub fn control_point(&self, loc: &Location)
+    {
+        // Since we don't hold an explicit lock, updating a Location is tricky: we might read a
+        // Location, work out what we'd like to update it to, and try updating it, only to find
+        // that another thread interrupted us part way through. We therefore use compare_and_swap
+        // to update values, allowing us to detect if we were interrupted. If we were interrupted,
+        // we simply retry the whole operation.
+        loop {
+            let pack = &loc.pack;
+            // We need Acquire ordering, as PHASE_COMPILED will need to read information written to
+            // external data as a result of the PHASE_TRACING -> PHASE_COMPILED transition.
+            let lp = pack.load(Ordering::Acquire);
+            match lp & PHASE_TAG {
+                PHASE_COUNTING => {
+                    let count = (lp & !PHASE_TAG) >> PHASE_TAG;
+                    let new_pack;
+                    if count >= self.hot_threshold.load(Ordering::Relaxed) {
+                        new_pack = PHASE_TRACING;
+                    } else {
+                        new_pack = PHASE_COUNTING | ((count + 1) << PHASE_TAG);
+                    }
+                    if pack.compare_and_swap(lp, new_pack, Ordering::Release) == lp {
+                        break;
```

Comment:
OK, so if I understand this correctly, this is a lock-free update to the pack. The atomics will:
 * inform the compiler to respect the ordering.
 * inform the hardware to respect the ordering by emitting fences into the compiled code.

The breaks from the loop took me a while to understand. We exit our loop if the old value was the same old value we observed before we tried to update the pack. If this was not the case, that suggests that another thread updated the pack in the meantime, and that we should try again.

Have I understood this correctly?

Did you base this section of code on an example you found somewhere on the internet? If so, could you share the link?

Reply:
Yes, that's a correct understanding of my intention. No, it's not based on other code.